### PR TITLE
Added tutorial for splitting factories into multiple files

### PR DESCRIPTION
**Pull request by rafal-radziszewski**
Hopefully solves #184.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+          body: "Still working on it!"
+        }
+      end
+
+    end
+  end
+end
+```
+
+That way you can split you giant factory file into many small files. But what about name conflicts? Use pattern matching to avoid them!
```
Comment (pdawczak) [inline README.md]: It looks like there is one extra "you":

> That way you can split **you** giant factory [...]

File: README.md [hunk 2]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
 used for different repos, your factory module is getting too big, or if you have
 different ways of saving the record for different types of factories.
 
+### How to split your factories into different files
```
Comment (pdawczak) [inline README.md]: Would

"Splitting factories into separate files"

sound better?

File: README.md [hunk 3]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+
+We will assume that you use factories only in test environment. If you use them in all environments, just adjust directory structure.
```
Comment (pdawczak) [inline README.md]: How about:

"This example shows how to set up factories for testing environment. For setting them in all environments, please see _To install in all environments_ section."

File: README.md [hunk 4]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+```elixir
+# test/support/factory.ex
+defmodule MyApp.Factory do
+  use ExMachina.Ecto, repo: MyApp.Repo
+
+  use MyApp.ArticleFactory
+end
+```
+
+`MyApp.ArticleFactory` will be the first factory we will create. Create separate directory for factories, like `test/factories`. Here is how to create a factory:
```
Comment (pdawczak) [inline README.md]: How about:

"Main factory includes `ArticleFactory`, let's create it next. It might be useful to keep the factories in separate directory, like `test/factories`:"

Comment (pdawczak) [inline README.md]: Quick question - does putting `*.ex` files in `test/factories` work _without_ updating [`elixirc_paths`](https://github.com/thoughtbot/ex_machina/pull/185/files#diff-04c6e90faac2675aa89e2176d2eec7d8R74)?

File: README.md [hunk 5]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+      def with_comments(%MyApp.Video{} = video) do
+        insert_pair(:comment, video: video)
+        video
+      end
+    end
+  end
+end
+```
+
+This way you can assure that you do not have any scoping problems with your methods.
```
Comment (rafal-radziszewski) [inline README.md]: The issue is about features explained elsewhere, mainly regarding creating models with assocs. If you start using a big number of methods creatings assocs (like `with_user`), you might encounter problems. Presented way of splitting is based on macros, so what we do is basically we put all those factories in one module - but we do it at compile time. While managing big number of factories in separate files is very convenient, you always need to remember, that this ultimately gets defined in one place. Since pattern matching is preferred solution to make sure, that you always use the right method. Otherwise you can encounter really weird bugs.

File: README.md (hunk 4 above)
Comment (rafal-radziszewski) [inline README.md]: It won't, I forgot to mention that. I will add a section regarding that.

File: README.md (hunk 2 above)
Comment (rafal-radziszewski) [inline README.md]: Yes, done.

File: README.md (hunk 3 above)
Comment (rafal-radziszewski) [inline README.md]: Done, though the section does not exists yet, suggestions what to put there are welcome.

File: README.md [hunk 6]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
 used for different repos, your factory module is getting too big, or if you have
 different ways of saving the record for different types of factories.
 
+### How to split your factories into different files
+
+We will assume that you use factories only in test environment. If you use them in all environments, just adjust directory structure.
+
+> Start by creating main factory module in `test/support/factory.ex` and name it `MyApp.Factory`.
```
Comment (rafal-radziszewski) [inline README.md]: Done.

File: README.md (hunk 4 above)
Comment (rafal-radziszewski) [inline README.md]: Done. Good point, it is clearer now.

File: README.md (hunk 3 above)
Comment (pdawczak) [inline README.md]: Oh sorry, the section I was referring to is the [one mentioned earlier in the file](https://github.com/thoughtbot/ex_machina/pull/185/files#diff-04c6e90faac2675aa89e2176d2eec7d8R57).

Hope that helps!

**Comment by paulcsmith** [comment]
@rafal-radziszewski Thank you so much for writing this. You are absolutely right that it is useful to have some documentation on splitting the factories up. My only concern is that putting the function definitions inside a `using` blog will make it hard to follow errors in stack traces since I believe the line numbers are muddied up. Have you had any issues with that in your application?

**Comment by paulcsmith** [comment]
And thanks @pdawczak for helping out with the review!

**Comment by rafal-radziszewski** [comment]
@paulcsmith Personally I have never encountered any hard-to-trace issues with factories. I chose that type of solution since I used something similar with Forge/Blacksmith (I think it was even mentioned in documentation). I have not come up with any better solution that would allow to use one main factory module, which is important because usually you do not want to maintain factories en masse and remember where something is defined (the same applies to navigating one huge file).

I have had a lot of work recently and  have quite forgotten about this PR. If you have any more issues, just leave a comment and I will fix it. I will also take another look at @pdawczak comments to see if I haven't missed something.

**Comment by paulcsmith** [comment]
@rafal-radziszewski Thanks so much. That is really helpful information. I'll go ahead and take a closer look on Friday and merge it in. If there are any changes required I'll go ahead and make them since you're busy. Good luck with what you're working on :D

**Comment by rafal-radziszewski** [comment]
@paulcsmith Actually I have much more time now :) Just ping me if you need anything. Great work on the library BTW, congrats!

File: README.md [hunk 7]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
 used for different repos, your factory module is getting too big, or if you have
 different ways of saving the record for different types of factories.
 
+### Splitting factories into separate files
+
+This example shows how to set up factories for testing environment. For setting them in all environments, please see _To install in all environments_ section
```
Comment (paulcsmith) [inline README.md]: A couple small changes here:

> for *the* testing environment

> please see *the* "To install in all environments" section"

File: README.md [hunk 8]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+
+> Start by creating main factory module in `test/support/factory.ex` and name it `MyApp.Factory`. The purpose of main factory is to allow you to include only single module in all tests.
```
Comment (paulcsmith) [inline README.md]: > of *the* main factory

> include only *a* single

File: README.md [hunk 9]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+```elixir
+# test/support/factory.ex
+defmodule MyApp.Factory do
+  use ExMachina.Ecto, repo: MyApp.Repo
+
+  use MyApp.ArticleFactory
+end
+```
+
+Main factory includes `MyApp.ArticleFactory`, so let's create it next. It might be useful to create separate directory for factories, like `test/factories`. Here is how to create a factory:
```
Comment (paulcsmith) [inline README.md]: > *The* main factory

> create *a* separate directory

File: README.md [hunk 10]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+defmodule MyApp.ArticleFactory do
+  defmacro __using__(_opts) do
+    quote do
+
+      def article_factory do
+        %MyApp.Article{
+          title: "My awesome article!",
+          body: "Still working on it!"
+        }
+      end
```
Comment (paulcsmith) [inline README.md]: Could you remove the whitespace above and below this function?

File: README.md [hunk 11]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+This example shows how to set up factories for testing environment. For setting them in all environments, please see _To install in all environments_ section
+
+> Start by creating main factory module in `test/support/factory.ex` and name it `MyApp.Factory`. The purpose of main factory is to allow you to include only single module in all tests.
+
+```elixir
+# test/support/factory.ex
+defmodule MyApp.Factory do
+  use ExMachina.Ecto, repo: MyApp.Repo
+
+  use MyApp.ArticleFactory
```
Comment (paulcsmith) [inline README.md]: These two `use`s could go together without the white space I think :)

File: README.md [hunk 12]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+          body: "Still working on it!"
+        }
+      end
+
+    end
+  end
+end
+```
+
+That way you can split your giant factory file into many small files. But what about name conflicts? Use pattern matching to avoid them!
```
Comment (paulcsmith) [inline README.md]: > That -> This

File: README.md [hunk 13]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+
+```elixir
+# test/factories/post_factory.ex
+defmodule MyApp.PostFactory do
+  defmacro __using__(_opts) do
+    quote do
+
```
Comment (paulcsmith) [inline README.md]: You can remove this newline

File: README.md [hunk 14]
```diff
@@ -185,6 +185,85 @@ the same directory. This can be helpful if you need to create factories that are
...
+      end
+    end
+  end
+end
+
+# test/factories/video_factory.ex
+defmodule MyApp.VideoFactory do
+  defmacro __using__(_opts) do
+    quote do
+
```
Comment (paulcsmith) [inline README.md]: And this newline

File: README.md (hunk 5 above)
Comment (paulcsmith) [inline README.md]: I think we can remove this line since the example does a pretty good job showing what to do. We can add it back in if people have trouble with it in the future

**Comment by rafal-radziszewski** [comment]
@paulcsmith Fixes done. Thanks for fixing my English ;) About whitespaces: I mostly separate the quote inside from the quote, since for me it improves readability. I adjusted them too as you requested though.
